readme.md

![GeoGab Logo](./images/GeoGab.svg)

#### Gabriel A. Sieben (Physicist)
Here is a little code I wrote for my HomeDev Smarthings projects. I thought it could be useful for others too. Have fun. Gabriel :)

# Runtime Meter V 1.0.5
The **RuntimeMeter** measures the runtime of certain sections of the code of a loop. The sections can be defined freely. This is used to estimate the utilization of the CPU. The word estimation means that there can never be a perfect measurement of the code runtime. (Special code excepted.). Only `if()` conditions cause slight runtime differences. The measurement is to be regarded therefore as exact, but not as precise. Because by nature there are runtime fluctuations in every loop pass. 

It is a widespread misconception that microprocessors are always busy. In fact, I argue that in most cases the CPU is far from being busy. It depends decisively on the code whether they are or not. 

## Tested On
ESP32, ESP8266

# Examples Printout Screenshot
<img src="./images/printout.png" alt="Print Out Example" height="300"/>

# Example ESP Webserver Screenshot
Here you can see the difference between an 'ESP32' and 'ESP8266'. Both run on 160 MHz. But the ESP32 shows almost no kernel runtime, because the second CPU does the kernel tasks. Especially you should keep an eye on the loop frequency! 

<img src="./images/ESP32.gif" alt="Webserver Example EPS32" height="300"/><img src="./images/ESP8266.gif" alt="Webserver Example EPS8266" height="300"/>

## The effective code
A `delay(x)` can mean an eternity of doing nothing for the processor. This is because modern microprocessors are fast. Even much faster than many assume. Except for `delay(0)`, which we will come back to later. A good programmer should generally do without `delay(x)` and rather trigger the code based on events. Like the `.print(x)` of this library.

A well written program will do without a `delay(x)`. The loop is run completely unbraked. This is sometimes called asynchronous. The code is event driven. Such an event can be, for example, that 1000 milliseconds have passed since the last execution of the code, and the code should be executed now. This is exactly what the `.print(x)` of this library does. It is executed only every `x` milliseconds. Otherwise it does nothing. In this case it should e.g. print the information for the developer on the serial port. 

But with even such a program it can happen that the CPU gets bored. It runs the loop constantly and does not wait. But if there is no event, it does not do anything else than wait for events in the loop. So strictly speaking, it is also a CPU waiting time. And that's where the *RuntimeMeter* comes in. 

## Preliminary to the understanding
At least the programmers of an EPS know. The kernel also needs computing time. So the base system that provides all the great features like WiFi, Serial Port etc.. And we have to take care of that. Because a microprocessor is not a multitasking system, where this happens automatically in the background. Dual processors like the ESP32 are included as a kind of special form. We will come back to this one later. The second core is still ignored in the current version of the *RuntimeMeter*s, but this does not play a role in the runtime determinations of the first core. 
So, as is clear to many, the kernel urgently needs a `yield()`, or `delay()`. But where, when and for how long is not clear to a lot of people. Some functions of the kernel are executed immediately. A `serial.print()` sends the data immediately into the buffer. But if the buffer is full, the `Serial.print()` will not return until the buffer has space again. The web server of the EPS has its own handle, for example. You have to call it specifically in the loop to make the webserver work. But the rest, is executed in the time of the `yield()` or `delay()`. 

There is a "special form" of `delay()` that hardly anyone knows. The `delay(0)`. During `delay(0)` the kernel does all basic tasks and returns without waiting after they are done. 

## The principle of the *RuntimeMeter*.
The *RuntimeMeter* brings light into the shadow. The principle looks like this. 

    loop {
        [start]
            Program code
        [Intermediate]
            Further program code
        [End]

    [Evaluation]
        The determined data are as follows:
        - Runtime between individual measuring points
        - Runtime of the kernel
    }

## What we learn from the acquired data
The *RuntimeMeter* can measure the runtime in processor cycles, micro seconds and milliseconds. The measurement in micro and milliseconds will not be discussed further here, since the principle is the same. They are less precise, but larger runtimes can be measured, which will hardly be needed in practice. More about this later. 

The RutimeMeter measures how many cycles are brought between the measuring point of a loop pass. In the above example there are three runtimes which are measured. Namely:

- [Runtime1]=[Start]-[Intermediate time]
- [Runtime2]=[Intermediate]-[End]
- [Kernel]=within the [End] the kernel runtime is determined. 

So the total runtime of the code inside a loop is `[Loop]=[Runtime1]+[Runtime2]+[Kernel]`. Any code outside of `[start]` and `[end]` is not considered. If possible, it should only be used for data processing, to falsify the measurement results as little as possible. 

The collected data offer us two rudimentary findings:

o In the direct comparison of the run times, one can quickly see which parts of the program take how long. If one changes the code of a part, this can directly affect its runtime.

o The frequency of the loop run. And this is especially important for the "without delay method" of programming. Because as said before: Running in the loop to wait for events is no real processor load. At least not as long as you could reach the same if you check the events less often. If the processor is more loaded, the loop frequency will automatically decrease. Which value is a good value, the programmer must decide for his application. E.g. there are applications where a LED display should be operated with 50Hz. If this is not reached, the *RuntimeMeter* will tell you immediately and you can make appropriate code changes. 

## The runtime fluctuations
As mentioned above, the runtime measurement is subject to slight fluctuations. Everyone can recognize this already alone by the example programs of this library. This is by no means a programming error of the *RuntimeMeter*. The measurements are accurate, but not precise. The accuracy meets the requirements of the *RuntimeMeter*s. It is used for estimation and not for exact measurement. A few cycles faster or slower certainly do not play a role in the flow of a complex program. If the code must be measured exactly, the *RuntimeMeter* is the wrong choice. You will have to translate the code into machine code. Then decompile it back to assembler, which then has to be studied in more detail. Mostly one does not write such programs therefore into a high-level language but directly into Assembler, which is converted mostly directly into machine code. But this shall not be discussed further here.

There are several reasons why runtime variations occur. In the core it is of course always the same reason. A foreign code is executed outside of the own program, which flows into the run time. This code is on the one hand the RutimeMeter Library itself. Even by tricks this cannot be avoided completely. So the first measuring point initializes the first measurement. That's all that needs to happen. Only the second measuring point determines the runtime of the first. Of course, one could further optimize the RutimeMeter at one or the other point. But you would not gain more knowledge. Because the *RuntimeMeter* has hardly any influence on the fluctuations. The far greater part comes from the kernel. As mentioned above, some system tasks are done immediately. Others wait for the `delay()`. 

To determine the runtime more exactly, brings with the small values of the fluctuations, no realization of use. 


## The implementation
## Limitations
### No precession measurement
This has been sufficiently discussed above. Keyword runtime fluctuations. 

### The length of the runtime measurement is limited
I consider the possible length of the transit time measurements to be sufficient. If the runtime in cycles is not sufficient, you can use the mircos and millis method, where microseconds and milliseconds are used instead of cycles. For such "long" runtimes it does not matter if they are still calculated in cycles. The limitation is due to a simple fact. I use "only" 32bit variables. I could also use 64Bit. But then they consume more memory and can't be processed as fast. I will not go into this now.

An unsigned 32 bit variable (uint32_t) can store number up to `2^32=4.294.967.296`. At `2^32` a rollover occurs. That means we would need one bit more (which does not exist). Therefore with the value `uint32_t x=2^32+1` the variable `x` starts again at zero. 

The *RuntimeMeter* counts everything with 32 bits. That means, in case of microseconds, it can store a maximum of 4.294.967.296 microseconds. And this restriction is valid for the single measurement as well as for the total runtime measurement. So all measured values added together must not overwrite 2^32 units (cycles, microseconds, milliseconds). 

In practice these are runtimes up to: 

**Milliseconds**: 1,000 milliseconds = 1 second. => `2^32/1000` seconds => 1 hour is known to be (60 seconds * 60 minutes) => Possible runtime: `2^32/1000/60/60` = 1.193 hours

**Microseconds**: 1,000 microseconds = 1 millisecond => 1,193 h / 1000 = 1.2 hours

**Nanoseconds**: Not useful! With today's µCPUs one 
nanosecond is usually more than 1 CPU tick. (e.g. at 160MHz one tick = 12.5 nanoseconds long).

**Picoseconds**: As above. 

**CPU cycles**: (recommended setting):
Here is a little more complex. For this you need to know what a tick is. The tick is a CPU clock unit. Usually a processor can do one calculation step per clock unit (tick). I will not go into more detail now. 

We know: 1 Second (s) = 1.000 [1E3] Milliseconds (ms) = 1.000.0000 [1E6] Microseconds (µs) = 1.000.000.000 [1E9] Nanoseconds (ns) = 1.000.000.000 [1E12] Picoseconds (ps)

We also know what frequency our CPU is running at. It is usually given in MHz (megahertz) for µCPUs. This is 1.000 kHz or 1.000.000 Hz (Hertz). 1 Hz is once per second. In our case one tick or cycle. In the case of the frequency of the loop, one round. This means that a CPU has the frequency times cycles per second. 

Let's consider an ESP8266 with 80 MHz. In one second, the processor can perform 80,000,000 arithmetic operations (I told you. They're fast!). Hertz (Hz) corresponds to the unit 1/sec. So 1/frequency equals the time for one tick. `1/80.000.000 Hz = 0.000.000.012.5 seconds`. So 12.5 nano seconds pass per tick. 

So with 2^32 cycles the time span can be 2^32*12.5 nano seconds:   

CPU Frequency: xxx MHz -> xxx cycles per µs =>.

[**80 MHz**]: ``2^32=4,294,967,296`` => ``4,294,967,296/80`` => ~ ``53,687,091 µs`` => ~ ``53,687 ms`` => ~ ``53 sec``

[**160 MHz**]: ``2^32=4,294,967,296`` => ``4,294,967,296/160`` => ~ ``26,843,545 µs`` => ~ ``26,843 ms`` => ~ ``26 sec``  

[**240 MHz**]: ``2^32=4,294,967,296`` => ``4,294,967,296/240`` => ~ ``7.895.697 µs`` => ~ ``17.895 ms`` => ~ ``17 sec``

Conclusion: Even the 17 seconds loop time for an ESP32 with 240 Hz is more than sufficient. Because 17 seconds is not a reasonable loop time anyway. With this, the system would certainly no longer work well. 


## Application
### For development
The software can be used for analysis purposes during development. If the program does not use the runtimes further, the code can be left in place, but it is not necessary. It can be removed after the development. This saves a little bit of resources. 

### For permanent performance evaluation
The software can be implemented permanently. The resources required are kept within limits. An example can be found in the examples. It is a quick and dirty example web server that lets you retrieve the processor load at any time. The software serves also programs, which permit the loading of strange program code. So the programmer of the foreign program code can observe the influence of his code. Especially for realtime applications this can be very interesting. E.g. the project WLED by Aircoookie , which I love very much. 

## Instruction
### Include
``#include <GeoGabRuntimeMeter.h>``

### Create object
`RuntimeMeter [object name] ([number of slots],[measurement type]);`

`RuntimeMeter rtmeter(); // 5 slots, RT_MEASURE_CYCLES by default`
            
`RuntimeMeter rtmeter(10, RT_MEASURE_MICROS); // 10 slots, RT_MEASURE_MICROS` 

Measurement Types are: cycles (ticks): ``RT_MEASURE_CYCLES`` (default) , micros seconds: ``RT_MEASURE_MICROS``, milli seconds: ``RT_MEASURE_MILLIS``

### Measurements
1. add measurement point:
    **Method**: ``Add();``
    **Parameter**: (string) name of the measurement point
    **Return**: (uint32_t) Runtime of the last measurement. 0 = Error;

2. end measurement: 
    **Method**: ``Finalize();`` 
    **Parameter**: (bool) true(default) = measure the kernel time, false: = no kernel time. Keep in mind, however, that then somewhere after the code must follow at least a delay(0);.
    **Return**: (char) error number, where 0 = no error, If you use Print(), you can safely ignore the error message. Print gives a hint. 

3. **Variant**:  ``rtmeter.add(); // measure point``

*or*

``rtmeter.add("name"); // measure point with name (for the following measurement), (default: number of the measurement)``

*or* 

``runtime=rtmeter.add("This measure point"); // return the runtime of the last measure point in uint32_t``

``rtmeter.finalize(); // This must always be the last measure point.``

**Special form**:
``rtmeter.pause(); // Pauses the measurement until the next Add(). The runtime in between is not added to the total runtime``.

### Evaluation
1. print the information for the developer on the Serial.port. 
     **Method**: print();    
    **Parameter**: (uint16_t) Waiting time until next print in milliseconds (default: 1,000).

2. direct access to the public variables of the *RuntimeMeter*:
    There are three groups of variables: 
      * Variables concerning the loop: rtmeter.loop
      * Variables concerning the measuring points: rtmeter.mpslots
      * Additional variables: rtmeter.overall


```cpp
rtmeter.mpslots[no].
      String Name                       // Slot Name
      uint32_t Runtime                  // Runtime
      uint16_t Percentage               // Percentage multiplied by 10 (Divide the number by 10 and you get the first decimal place)
```

```cpp
    rtmeter.overall.
      uint32_t LoopCounter              // Counts the number of loops since program start (no rollover handling)
      uint32_t LastPrintMillis          // Time index at last print
      measurePoint_t kernel             // Kernel (same structure as mpslots) [e.g. rtmeter.overall.kernel.Runtime]
      uint16_t cpuf                     // CPU Frequency
```

```cpp
    rtmeter.loop.
      uint8_t Error                     // Error Number
      uint32_t PrevStamp                // Previous Time Stamp (internal)
      uint32_t ActStamp                 // Actual Time Stamp (internal)
      uint32_t Runtime                  // Loop Runtime
      uint32_t Frequence                // Loop Frequency Hertz multiplied by 10 (Divide the number by 10 and you get the first decimal place) 
      uint16_t SlotIndex                // Index of the actual slot+1 (good for loops)
      bool Finalized                    // Will be set to true once the loop is finalzed (internal)``
```

e.g.

``rtmeter.loop.Runtime; // Overall Runtime``

``(int) rtmeter.loop.Frequence/10;         // Frequency of the loop in Hertz (1/Sekunde) without decimal place``
``(float) rtmeter.loop.Frequence/10;       // Same with decimal place``

``rtmeter.overall.kernel.Runtime``
``rtmeter.mpslots[1].Name                  // Measure Point 2``
``rtmeter.mpslots[0].Runtime               // Measure Point 1``

etc...

### Error codes
The error adds up: there are the following errors:

**0**: 0000 0000: Perfect :-)           

**1**: 0000 0001: More measurement points than slots. Please reserve more slots (*RuntimeMeter* rtmeter(xxx);)   
**2**: 0000 0010: Nothing was measured. You need at least one measurement slot. Add(); and the Finalize();

**4**: 0000 0100: Before the Finalize, Print() makes no sense, because the data has not yet been collected completely. 

**8**: 0000 1000: The collection mode is unknown. Please choose one of the following values: RT_MEASURE_CYCLES, RT_MEASURE_MICROS, RT_MEASURE_MILLIS, RT_MEASURE_SECS

Example: Error *9* (*1+8*) => 0000 1001

### Example

```csharp
RuntimeMeter rtmeter(10); // Allows to measure 10 Points 

loop {
  rtmeter.add("Measurement 1"); // Starts the measurement with a name
    /*... this code will be measured */

  rtmeter.add("Measurement 2"); // Sets a measurement point with a name
    /*... this code will be measured */

  uint16 last runtime=rtmeter.add("Measurement 3"); // Sets a measurement point with a name and writes the last runtime into the variable last

  rtmeter.pause(); // Pauses the measurement // uint16 last rtmeter.pause(); 

    /*... This code is not measured. This time frame is also not included in the total runtime */.

  rtmeter.add(); // Starts the measurement again without name

   /* ... this code will be measured */

  rmeter.finalize(); // Finishes the last measurement, measures the kernel runtime, calculates the percentages 

  rmeter.print(1000); // prints the values 1x in the second.
}
```

# Installation
Download the source of my library and unpack it into your libs folder. Include the library with `#define <GeoGabRuntimeMeter.h>`. 

Or use the library dependencies in Visual Studio Code including PlatformIO:

lib_deps =
      gsieben/GeoGab-RuntimeMeter @ ~1.0.5


# License information
The software is subject to the MIT license model. It is relatively free to use. See license file.



changes.md
# Future Release
- Seaching for the support of Arduino ATmega328P. Can't find a way to get the cpu cycles. All other problems can be fixed. 

# Version 1.0.3
- Added ESP32 support to the "EPS Webserver Example"
- Small changes in the readme.md
- Small changes in the example "EPS Webserver"
- Bug Fix: No Name once .Add() was empty.

# Version 1.0.2
- Minor Bug Fixes

# Version 1.0.1
- Minor Bug Fixes

# Version 1.0.0
- First Release



src/GeoGabRuntimeMeter.h
/**
 * @file RuntimeMeter.h
 * @author Gabriel Sieben (gsieben@geogab.com)
 * @brief 
 * @version 1.0.5
 * @date 2021-02-11
 * 
 * @copyright Copyright (c) 2021
 * 
 * @licenze MIT license (see LICENSE file)
 */

#ifndef GeoGabRuntimeMeter_h
#define GeoGabRuntimeMeter_h

#include <Arduino.h>

#define XSTR(s) STR(s)
#define STR(...) #__VA_ARGS__

#define GEOGAB_RM_VERSION "1.0.5"

enum  {
  RT_ERROR_NONE=0,                        // 0000 0000: Perfect :-)           
  RT_ERROR_NO_SLOTS_FREE=1,               // 0000 0001: Reverse more slots on init    
  RT_ERROR_NO_DATA_CAPTURED=2,            // 0000 0010: Set at least one capture points
  RT_ERROR_FINALIZE_BEFORE_PRINT=4,       // 0000 0100: Print does not make sence before consolidation 
  RT_ERROR_UNKNOWN_MODE=8,                // 0000 1000: Der Erfassungsmodus ist unbekannt. Bitte wähle einen folgender Werte: RT_MEASURE_TICKS, RT_MEASURE_MICROS, RT_MEASURE_MILLIS, RT_MEASURE_SECS
};

enum {
  RT_MEASURE_CYCLES,
  RT_MEASURE_MICROS,
  RT_MEASURE_MILLIS
};


class RuntimeMeter {
  public:
    /* Constructors */
    RuntimeMeter ();
    RuntimeMeter (uint8_t size_);  // Pups pups pups
    RuntimeMeter (uint8_t size_, uint8_t Mode_);    // Dadada

    
    /* Prototypes */
    uint32_t Add(const String &name="");
    uint32_t Pause();  
    bool Finalize(bool kernel=true);
    void Print(const uint16_t &every=1000);   
    void FlushData();    
    String Version();

    /* Structures */
    struct measurePoint_t {
      String Name="";                   // Slot Name
      uint32_t Runtime=0;               // Runtime
      uint16_t Percentage=0;            // Percentage multiplied by 10 (Divide the number by 10 and you get the first decimal place)
    };

    struct overall_t {
      uint32_t LoopCounter=0;           // Counts the number of loops since program start
      uint32_t LastPrintMillis=0;       // Time index at last print
      measurePoint_t kernel;            // Messpunkt für den Kernel
      uint16_t cpuf=F_CPU/1000000;      // CPU Frequency
    };

    struct loop_t {
      uint8_t Error=0;                  // Number of the last error
      uint32_t PrevStamp=0;             // Previous Time Stamp
      uint32_t ActStamp=0;              // Actual Time Stamp
      uint32_t Runtime=0;               // Runtime in ticks
      uint32_t Frequence=0;             // Frequence Herz multiplied by 10 (Divide the number by 10 and you get the first decimal place) 
      uint16_t SlotIndex=0;             // Index of the actual slot
      bool Finalized=false;             // Will be set to true once the loop is finalzed
    };

    /* Variables */
    overall_t overall;
    loop_t loop;
    std::vector<measurePoint_t> mpslots;  // (Init 1) Vector of the current measured values

  private:
    /* Internal Prototypes */
    void CalcRuntime(uint32_t &target);
    void GetStamp(uint32_t &var);
    void Init();

      /* Internal Variables init be object creation */
    uint16_t NoOfSlots;   // (Init 2) Number of measuring points available (Default: 5)
    uint8_t Mode;         // (Init 3) Measuremnt Type (Default: RT_MEASURE_TICK=0)

    /* Internal Variables */
    uint32_t ActStamp=0;
    uint32_t PrevStamp=0;
    uint32_t OverallLoopRuntime=0;

    String ModeName[3] = {"CPU Ticks", "Micro Seconds", "Milli Seconds"}; 
    String ModeUnit[3] = {"", "µs", "ms"};
};

#endif


examples/PrintSerial/SerialPrintExample.cpp
/**
 * @file RuntimeMeterExample.cpp
 * @author Gabriel Sieben (gsieben@geogab.net)
 * @brief A simple example with a serial output of the result
 * @version 0.1
 * @date 2021-02-14
 * 
 * @copyright Copyright (c) 2021
 * 
 * @licenze MIT license (see LICENSE file)
 */
#include "SerialPrintExample.h"


  //RuntimeMeter rtmeter(3,RT_MEASURE_MICROS);          // 3 Slots but measure in micro seconds instead of cpu ticks (RT_MEASURE_TICKS) which is default. This is less accurate but longer time intervals can be measured.
  RuntimeMeter rtmeter(5);

/****** Setup ******/
void setup() {
  Serial.begin(115200);
  Serial.println("\r\n### S T A R T ###");
}


/****** Main Loop ******/
void loop() {
rtmeter.Add("Code A");          // Measurement Point 1 (Start)
  uint32_t b;
    for (uint32_t a=0;a<1000;a++) {
      b=acos(a);
    } 


rtmeter.Add("Code B");          // Measurement Point 2 (caculate runtime between Point1 and Point2)
    for (uint32_t a=0;a<100;a++) {
      b=acos(a);
    } 

rtmeter.Add("Code C");          // Measurement Point 2 (caculate runtime between Point1 and Point2)
    for (uint32_t a=0;a<100;a++) {
      b=acos(a);
    } 

rtmeter.Finalize();           // Last Measure Point and finalize calculations. The final() is mandatory!!! 

  /** Note:
   * 'Finalize()' also calculates the kernel time. These include WiFi, Flash, etc. In the zone of the time measurement a delay makes little sense. From here one 
   * can use a 'delay()' if necessary. Considered however, that each 'delay()' affects the measurement of the kernel further. Because within a 'delay()' the kernel 
   * tasks are processed. 
   */

  /** Note: 
   * As you will see, the runtimes are not always 100% equal. But the precession of the values should satisfy the requirements of an estimation. The values flucuate 
   * slightly, because system accesses can have different runtimes. This cannot be influenced.   
   */
 
  /** Note:
   * From 'Finalize()' on, all determined data can be read directly from the public variables for this purpose. The next 'Add()' destroys the pre-arranged data.
   *
   * There are the following groups of variables: rtmeter.loop,rtmeter.mpslots, rtmeter.overall
   * The access is direct: rtmeter.overall.kernel.Runtime, rtmeter.mpslots[1].Name, rtmeter.mpslots[1].Runtime
   */

  rtmeter.Print(2000);        // Optionally you can output the data on the serial port. By default, the data is output every second (1000 ms).

  rtmeter.loop.Runtime=b;     // This seems to not make any sense, but it serves a purpose in this example code. The purpose of this line is only to make the optimizer think the above for loops are not meaningless. This variable will be deleted in the next step anyway. 
}



src/GeoGabRuntimeMeter.cpp
/**
 * @file RuntimeMeter.cpp
 * @author Gabriel Sieben (gsieben@geogab.net)
 * @brief Runtime... Testes on ESP8266
 * 
 * @version 1.0.5
 * @date 2021-02-11
 * 
 * @copyright Copyright (c) 2021
 * 
 * @licenze MIT license (see LICENSE file)
 * 
 */
#include <GeoGabRuntimeMeter.h>

/**
 * @brief Runtime Meter: Measures the time intervals between markers.
 * 
 */
RuntimeMeter::RuntimeMeter () :  mpslots (5), NoOfSlots (5), Mode (RT_MEASURE_CYCLES) {Init();} 
/**
 * @brief Runtime Meter: Measures the time intervals between markers.
 * 
 * @param size_ Set the Number of measurepoints you whant to measure. 
 */
RuntimeMeter::RuntimeMeter (uint8_t size_)  : mpslots (size_), NoOfSlots (size_), Mode (RT_MEASURE_CYCLES) {Init();} 
/**
 * @brief Runtime Meter: Measures the time intervals between markers.
 * 
 * @param size_ Set the Number of measurepoints you whant to measure. 
 * @param Mode_ Select a measure mode: In Ticks: RT_MEASURE_CYCLES, In Micros Secs: RT_MEASURE_MICROS, In Milli Secs: RT_MEASURE_MILLIS
 */
RuntimeMeter::RuntimeMeter (uint8_t size_, uint8_t Mode_)  :  mpslots (size_), NoOfSlots (size_), Mode (Mode_) {Init();} 

void RuntimeMeter::Init(){
    overall.LastPrintMillis=millis();                // Initial Time Shot
}

/**
 * @brief Adds a measuring point | On the first call it will inizialize 
 * 
 * @param String Name of the Measuring point (Optional)
 * @return uint32_t Runtime of the last run, 0=Error (except on the first measurement point)
 */
uint32_t RuntimeMeter::Add(const String &name) {

    if (loop.Finalized) {
        FlushData();                                // Flush old Data on a new round
    } 

    if (loop.SlotIndex==NoOfSlots) {
        loop.Error+=RT_ERROR_NO_SLOTS_FREE;         // Add Error Flag
        return 0;
    }

    mpslots[loop.SlotIndex].Name=name;              // Set the name

    if (!loop.PrevStamp) {   
        loop.SlotIndex++;                           // Rise Slot Index
        GetStamp(loop.PrevStamp);                   // Initial Measurement
        return 0;
    } 

    GetStamp(loop.ActStamp);                        // Store actual timestamp
    CalcRuntime(mpslots[loop.SlotIndex-1].Runtime); // Calculate the runtime
    loop.SlotIndex++;                               // Rise Slot Index
    loop.PrevStamp=loop.ActStamp;                   // Save the stamp to previous

    return mpslots[loop.SlotIndex-2].Runtime; 
}

/**
 * @brief Pauses the measurement until the next add.  
 * 
 * @return uint32_t: Runtime of the last measurement
 */
uint32_t RuntimeMeter::Pause() {
    if (!loop.PrevStamp) {                            // If someone placed pause at the beginning or one pause after another -> do nothing
        return 0;
    } 

    GetStamp(loop.ActStamp);
    CalcRuntime(mpslots[loop.SlotIndex-1].Runtime);   // Caculate runtime (writes in actual slot using acttamp and revtamp)
    loop.PrevStamp=0;                                 // Forces next add to start or ignore pause if it is called again

    return mpslots[loop.SlotIndex-1].Runtime;
}


/**
 * @brief 
 * 
 * @param kernel Optional: If 'true' (default), the kernel runtime is calculated. If 'false', then at least a delay(0) after the finalzie until loopend is mandatory.
 * @return The runtime of the loop. 0=error. 
 */
bool RuntimeMeter::Finalize(bool kernel) {
    overall.LoopCounter++;                               // Just count the laps. Overflows at some point. But this does not matter. Serves only the development.
    if(loop.SlotIndex==0) {
        // No measurement no result. What a surprise ;-)
        loop.Error+=RT_ERROR_NO_DATA_CAPTURED;
        return  loop.Error;
    }

    /* Calculate the last runtime (if there was no pause before) */
    if(!loop.PrevStamp==0) {                             // true: Es muss noch eine Runtime berechnet werden
        GetStamp(loop.ActStamp);
        CalcRuntime(mpslots[loop.SlotIndex-1].Runtime);   // Caculate runtime (writes in actual slot using acttamp and revtamp)
    }


    /* Calculate the Kernel Runtime */
    if (kernel) {
        GetStamp(loop.PrevStamp);
        delay(0);                                       // This is where the magic is. delay(0) returns immediately after the kernel tasks are done.
        GetStamp(loop.ActStamp);
        CalcRuntime(overall.kernel.Runtime);            // Caculate runtime (using loop.ActStamp, loop.PrevStamp)
    }


    /* Overall Runtime */
    for (uint8_t i=0;i<loop.SlotIndex;i++) {            // ! No rollover check: maybe in the future i will do so error handling...
        loop.Runtime+= mpslots[i].Runtime; 
    }
    if (kernel) {
        loop.Runtime+=overall.kernel.Runtime;           // Add Kernel Runtime
    }


    /* Percentage multiplied by 10*/
    for (uint8_t i=0;i<loop.SlotIndex;i++) {
        mpslots[i].Percentage=(mpslots[i].Runtime*100*10/loop.Runtime);   // percentage=part/all*100
    }
    if (kernel) {
        overall.kernel.Percentage=(overall.kernel.Runtime*100*10/loop.Runtime);     // Calculate Kernel Part
    }

    /* Loop Frequence in Herz (1/sec) multiplied by 10  */
    switch ( Mode ){
        case RT_MEASURE_CYCLES :
        loop.Frequence=F_CPU/loop.Runtime;               // CPU Clock = Ticks per seccond e.g. [80 HHz] = 80.000.000 Ticks per Second
        break;
        case RT_MEASURE_MICROS :
        loop.Frequence=1000000*10/loop.Runtime;          // 1 sec = 1.000.000 micro seconds 
        break;
        case RT_MEASURE_MILLIS:
        loop.Frequence=1000*10/loop.Runtime;             // 1 sec = 1.000 milli seconds 
        break;
        default:
        loop.Error+=RT_ERROR_UNKNOWN_MODE;
    }
    loop.Finalized=true;
    return 0;
}


/**
 * @brief Outputs the results on the serial port for development purposes.
 * 
 * @param every Specifies how long (default 1000 ms=1 second) to wait until the next output. The specification is made in milli seconds. (Same as delay())
 */
void RuntimeMeter::Print(const uint16_t &every) {

    if (millis()<overall.LastPrintMillis+every) return;       // Der Rollover wird ignoriert. It only happens after 1.193 hours anyway. That doesn't matter for a development printout. (2^32=4294967296) Milli Secs -> 2^32/1000 Secs -> 2^32/1000/60/80 Hours
    overall.LastPrintMillis=millis(); 

    if (loop.Error) {
        Serial.printf("There is a previous error with the Number: %u\r\n", loop.Error);
        return;
    } 

    if (!loop.Runtime) {
        Serial.println(F("Please perform a consolidation before printing."));
        loop.Error+=RT_ERROR_FINALIZE_BEFORE_PRINT;
        return;
    } 
    
    Serial.println(F("RUNTIME METER SUMMARY:"));
    float nanospt=(float) 1000/overall.cpuf;
    Serial.printf("CPU Frequency: \t\t%u MHz\r\n", overall.cpuf);
    Serial.printf("Nano Seconds/Tick: \t%f\r\n",nanospt);
    Serial.printf("Slots Available:\t%u\r\n", NoOfSlots);
    Serial.printf("Measurements: \t\t%u\r\n", loop.SlotIndex);
    Serial.printf("Measure Mode: \t\t%s\r\n", ModeName[Mode].c_str());
    Serial.printf("Overall Loops: \t\t%u\r\n",overall.LoopCounter);
    Serial.printf("Loop Frequency: \t%u Hz (%u KHz)\r\n", loop.Frequence/10, loop.Frequence/10000);
    Serial.println(F(" ======================================================="));
    Serial.println(F("|Measurement\t|\tRuntime\t\t|\tShare\t|"));
    Serial.println(F("|===============|=======================|===============|"));
    for (uint8_t i=0;i<loop.SlotIndex;i++) {
        if(sizeof(mpslots[i].Name)) {
            //Serial.print(sizeof(mpslots[i].Name));
            Serial.printf("|%s\t\t",mpslots[i].Name.c_str());
        } else {
            Serial.printf("|%u\t\t",i+1);
        }
        Serial.printf("|\t%d %s\t\t|\t%u %%\t|\r\n",mpslots[i].Runtime, ModeUnit[Mode].c_str() ,mpslots[i].Percentage/10);
    }
    if(overall.kernel.Runtime) Serial.printf("|Kernel\t\t|\t%d %s \t\t|\t%u %%\t|\r\n", overall.kernel.Runtime, ModeUnit[Mode].c_str() , overall.kernel.Percentage/10);
    Serial.println(F("|---------------|-----------------------|---------------|"));
    Serial.printf("|Overall\t|\t%u %s\t\t|\t100 %%\t|\r\n",loop.Runtime, ModeUnit[Mode].c_str());
    Serial.println(F(" ======================================================="));
}

/**
* @brief Internal Function used to flush all Data / used by Finalized()
* 
*/
void RuntimeMeter::FlushData() {
    loop.Error=0;        
    loop.PrevStamp=0;    
    loop.ActStamp=0;     
    loop.Runtime=0;      
    loop.Frequence=0;
    loop.SlotIndex=0;   
    loop.Finalized=false;
}

/**
 * @brief Return the Integrator Version
 */
String RuntimeMeter::Version(){
  return GEOGAB_RM_VERSION;
}

/* P R I V A T E  F U C T I O N S */
/**
 * @brief Internal Function used to calculate the Runtime.
 * 
 * @param target Write directly to the target without copying the value (Reference)
 */
void RuntimeMeter::CalcRuntime(uint32_t &target) { 
    /* Calculate the runtime */
    if (loop.PrevStamp>loop.ActStamp) {                             // True if rollover happened | Note: Comparence valid for: [CPU 160 MHz-> Timedifference Max. 13 Seconds] -> See Nerd Knwoledge in readme.
        target = loop.ActStamp + (0xFFFFFFFF - loop.PrevStamp);     // Part1: 0 - ActSamp | Part2: PrevStam - End of Variable 
    } else {
        target = loop.ActStamp - loop.PrevStamp;                    // without overfow (rollover) the calculation is trivial
    }
}

/**
 * @brief Internal Function: Used to get the time based on the user selection
 * 
 * @param var Write directly to the target without copying the value (Reference)
 */
void RuntimeMeter::GetStamp(uint32_t &var) {
    switch ( Mode ){
         case RT_MEASURE_CYCLES :
            var=ESP.getCycleCount();
            break;
         case RT_MEASURE_MICROS :
            var=micros();
            break;
         case RT_MEASURE_MILLIS:
            var=millis();
            break;
         default:
            loop.Error=RT_ERROR_UNKNOWN_MODE;
    }
}
